### PR TITLE
array handling improvements

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
This PR fixes array handling with JDBC connections, which was missing the translation of these array types into the JDBC metadata, as well as the implicit array to JSON string conversion we were doing is not suitable for this use, since avatica translates them into `java.sql.Array`.

This PR also splits up the SQL array functions (prefixed `ARRAY_`) from multi value string functions (prefix `MV_`) to allow for subtly different return type reporting and handling (the primary difference being that `MV_` functions which return arrays will typically report their output type as `VARCHAR` instead of an `ARRAY`). I've also homogenized the type signatures of these functions to make sure all of the `ARRAY_` and `MV_` functions are consistent within the family (some of the ARRAY_ methods were missing handling actual ARRAY typed inputs heh).

To go along with this split, I also split the multi-value string and array unit tests into their own test files, since `CalciteQueryTest` is a monster, as well as adding a few additional tests, which also uncovered a bug in string array to numeric array conversion with null values.

<hr>

This PR has:
- [x] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Review by jihoonson — APPROVED** [review]
LGTM

File: docs/querying/sql.md [hunk 1]
```diff
@@ -1022,6 +1022,7 @@ Connection context can be specified as JDBC connection properties or as a "conte
 |---------|-----------|-------------|
 |`sqlQueryId`|Unique identifier given to this SQL query. For HTTP client, it will be returned in `X-Druid-SQL-Query-Id` header.|auto-generated|
 |`sqlTimeZone`|Sets the time zone for this connection, which will affect how time functions and timestamp literals behave. Should be a time zone name like "America/Los_Angeles" or offset like "-08:00".|druid.sql.planner.sqlTimeZone on the Broker (default: UTC)|
+|`sqlStringifyArrays`|When set to true, result columns which return array values will be serialized into a JSON string in the response instead of as an array (default: true, except for JDBC connections, where it is always false)|
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply docs/querying/sql.md:1025]: Since there are no other changes I'll try to do this in a follow-up to not churn through CI again

**Comment by clintropolis** [comment]
thanks for review @jihoonson

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/ArrayOffsetOperatorConversion.java [hunk 2]
```diff
@@ -39,16 +38,15 @@
...
-            OperandTypes.or(
-                OperandTypes.family(SqlTypeFamily.STRING),
-                OperandTypes.family(SqlTypeFamily.ARRAY),
-                OperandTypes.family(SqlTypeFamily.MULTISET)
-            ),
+              OperandTypes.or(
+                  OperandTypes.family(SqlTypeFamily.ARRAY),
+                  OperandTypes.family(SqlTypeFamily.STRING)
+              ),
             OperandTypes.family(SqlTypeFamily.NUMERIC)
```
Comment (rohangarg) [inline sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/ArrayOffsetOperatorConversion.java:45]: should be stricter - like having `SqlTypeFamily.INTEGER`?

Comment (clintropolis) [reply sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/ArrayOffsetOperatorConversion.java:45]: ah yeah, that probably could stand to be a bit further restricted

**Pull request merged**